src/history/hs_rec.c

```diff
@@ -492,7 +492,8 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_MULTI *mult
              * If we've reached a full update and it's in the history store we don't need to
              * continue as anything beyond this point won't help with calculating deltas.
              */
-            if (upd->type == WT_UPDATE_STANDARD && F_ISSET(upd, WT_UPDATE_HS))
+            if (upd->type == WT_UPDATE_STANDARD &&
+              F_ISSET(upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS))
```

Comment:
As we are not inserting the same update again into the history store because it already exists can this lead to a situation where the history store update doesn't have a stop timestamp?

Prepare update P1 -> U1.
Reconcile leads P1 to the data store and U1 to history store with MAX stop timestamp.
Rollback prepare transaction leaves the U1 in the HS.
Insert another update U2
Reconciliation leaves U1 as it is.

Are we handling this scenario somewhere?

Reply:
Yes, we are. Now rec_upd_select will walk the whole tree to delete any update with the WT_UPDATE_TO_DELETE_FROM_HS flag. Actually in that case, there is no need to check this flag in hs code.

Reply:
I think we should see this flag in reconciliation. If checkpoint writes the restored update to the data store, we should not see this in hs code. If it keeps the prepared update, it should not even save the update chain so we can't be here. If it writes a newer update, checkpoint should have removed it from hs in rec_upd_select.